### PR TITLE
Rework cinterop linkerOpts/compilerOpts (KT-29970)

### PR DESCRIPTION
**Pull request by LepilkinaElena**

### CODE REVIEW

**Comment by LepilkinaElena** [comment]
Fix for https://youtrack.jetbrains.com/issue/KT-29970

Other parts of change in other repos:

- https://github.com/JetBrains/kotlin/pull/2206
- https://github.com/JetBrains/kotlin-native-shared/pull/5

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt [hunk 1]
```diff
@@ -59,10 +59,20 @@ fun getCInteropArguments(): List<OptionDescriptor> {
...
                     "header file to produce kotlin bindings for", isMultiple = true, delimiter = ","),
-            OptionDescriptor(ArgType.String(), "compilerOpts", "copt",
-                    "additional compiler options", isMultiple = true, delimiter = " "),
-            OptionDescriptor(ArgType.String(), "linkerOpts", "lopt",
-                    "additional linker options", isMultiple = true, delimiter = " "),
+            OptionDescriptor(ArgType.String(), "compilerOpts",
+                    description = "additional compiler options (allow add several options divided by spaces)",
+                    isMultiple = true, delimiter = " "),
+            OptionDescriptor(ArgType.String(), "linkerOpts",
+                    description = "additional linker options (allow add several options divided by spaces)",
```
Comment (olonho) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt]: "allows to add several options separated by spaces"

File: backend.native/tests/build.gradle [hunk 2]
```diff
@@ -2879,6 +2879,20 @@ task inline_defaultArgs_linkTest(type: LinkKonanTest) {
     lib = "codegen/inline/defaultArgs_linkTest_lib.kt"
 }
 
+def generateWithSpaceDefFile() {
```
Comment (ilmat192) [inline backend.native/tests/build.gradle]: This generated file is treated by Git as a new one untracked file. So I think it's better to generate it in the `build` directory which is ignored by Git.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt [hunk 3]
```diff
@@ -59,10 +59,20 @@ fun getCInteropArguments(): List<OptionDescriptor> {
                     isMultiple = true, delimiter = ",", deprecatedWarning = "Option -h is deprecated. Please use -header."),
             OptionDescriptor(ArgType.String(), HEADER_FILTER_ADDITIONAL_SEARCH_PREFIX, "hfasp",
                     "header file to produce kotlin bindings for", isMultiple = true, delimiter = ","),
-            OptionDescriptor(ArgType.String(), "compilerOpts", "copt",
-                    "additional compiler options", isMultiple = true, delimiter = " "),
-            OptionDescriptor(ArgType.String(), "linkerOpts", "lopt",
-                    "additional linker options", isMultiple = true, delimiter = " "),
+            OptionDescriptor(ArgType.String(), "compilerOpts",
+                    description = "additional compiler options (allows to add several options separated by spaces)",
```
Comment (olonho) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt]: Not in this PR, but may make sense to add options matching to compiler ones, i.e. `-compiler-options`.

Comment (LepilkinaElena) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt]: Make these  deprecated and add new the same? Then CompilerOpt and linkerOpt for single value can be deleted6 they aren't released and nobody use them?

File: backend.native/tests/build.gradle [hunk 4]
```diff
@@ -2879,6 +2879,20 @@ task inline_defaultArgs_linkTest(type: LinkKonanTest) {
...
+    file.write """
+headers = stdio.h "${projectDir}/interop/basics/custom headers/custom.h"
+
+---
+
+int customCompare(const char* str1, const char* str2) {
+    return custom_strcmp(str1, str2);
+}
+"""
+    return "${buildDir}/withSpaces.def"
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/build.gradle]: `file.absolutePath`?

File: tools/kotlin-native-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/konan/KonanPlugin.kt [hunk 5]
```diff
@@ -206,6 +206,7 @@ internal fun MutableList<String>.addFileArgs(parameter: String, values: Collecti
 internal fun MutableList<String>.addListArg(parameter: String, values: List<String>) {
     if (values.isNotEmpty()) {
         addArg(parameter, values.joinToString(separator = " "))
+        println("$parameter, ${values.map { "\"$it\"" }.joinToString(separator = " ")}")
```
Comment (SvyatoslavScherbina) [inline tools/kotlin-native-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/konan/KonanPlugin.kt]: Debug output?

Comment (LepilkinaElena) [reply tools/kotlin-native-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/konan/KonanPlugin.kt]: Yes, thank you. This function now isn't used at all. Is it better to remove it?

**Comment by SvyatoslavScherbina** [comment]
Linker options from .def file are copied to library manifest.
Does library reader parse quoted linker options properly?

File: tools/kotlin-native-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/konan/KonanPlugin.kt (hunk 5 above)
Comment (SvyatoslavScherbina) [reply tools/kotlin-native-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/konan/KonanPlugin.kt]: Yes, why not?

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt (hunk 3 above)
Comment (olonho) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/CommandLine.kt]: Yeah.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt [hunk 6]
```diff
@@ -15,7 +15,3 @@ class KonanDeclarationTable(builtIns: IrBuiltIns, descriptorTable: DescriptorTab
         loadKnownBuiltins()
     }
 }
-
-// This is what we pre-populate tables with.
-val IrBuiltIns.knownBuiltins
```
Comment (LepilkinaElena) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt]: @alexander-gorshenev As I understood this is already on common part of serialization in Kotlin.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/ir/util/IrUnboundSymbolReplacer.kt [hunk 7]
```diff
@@ -19,36 +19,6 @@ import org.jetbrains.kotlin.ir.symbols.*
 import org.jetbrains.kotlin.ir.types.toKotlinType
 import org.jetbrains.kotlin.ir.visitors.*
 
-@Deprecated("")
-internal fun IrModuleFragment.replaceUnboundSymbols(context: Context) {
```
Comment (LepilkinaElena) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/ir/util/IrUnboundSymbolReplacer.kt]: @alexander-gorshenev JS team remove such function and there is no call for it now, so also remove.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt (hunk 6 above)
Comment (alexander-gorshenev) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt]: yep.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt [hunk 8]
```diff
@@ -219,8 +219,10 @@ internal tailrec fun IrDeclaration.findPackage(): IrPackageFragment {
             ?: (parent as IrDeclaration).findPackage()
 }
 
-fun IrFunction.isComparisonFunction(map: Map<SimpleType, IrSimpleFunctionSymbol>): Boolean =
-        this.symbol in map.values
+fun IrDeclaration.isIrBuiltIn() = this.findPackage().packageFragmentDescriptor is IrBuiltinsPackageFragmentDescriptor
+
+fun IrFunction.isComparisonFunction(map: Map<SimpleType, IrSimpleFunctionSymbol>) =
+        this.symbol.descriptor in map.values.map {it.descriptor}
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: What is this required for?

Comment (LepilkinaElena) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: It's used in evaluateOperatorCall https://github.com/JetBrains/kotlin-native/blob/b19d02885d89aeaaf8d7fd2d783eb24ed993ef1a/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt#L2107

I just removed extra return type.

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: `this in map.values` is clearly different from `this.symbol.descriptor in map.values.map {it.descriptor}`.

Comment (LepilkinaElena) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: Yes, I understood, I commited experiments when was debugging. Fixed as I made before.

**Pull request merged**